### PR TITLE
Announce

### PR DESCRIPTION
**Pull request by atomgardner**
Here's some alpha quality work towards #255 

Please take note of the "popular repository" indelibly front and centre on my Microsoft Github profile page.

### CODE REVIEW

**Comment by casey** [comment]
Currently every top-level item is in its own file, and I'd like to keep doing that. It use [fzf-vim](https://github.com/junegunn/fzf.vim) to navigate around the project, and it makes it very easy to jump between structs by filename, and not have to search within a file.

What do you think about creating a `tracker` submodule in `tracker.rs`, and then renaming `UdpTrackerConn` to `Client` and putting it into `src/tracker/client.rs`? There might be an HTTP tracker in the future, or there might not be, so it seems unnecessary to plan in advance for it.

It makes sense to put request/response pairs in the the same file, so, for example, `AnnounceRequest` and `AnnounceResponse` can both go in `src/tracker/announce.rs`.

`State` can go in `src/tracker/state.rs`.

More comments in a second, but I wanted to start with that high level thought.

File: src/error.rs [hunk 1]
```diff
@@ -167,12 +181,24 @@ impl From<globset::Error> for Error {
   }
 }
 
+impl From<host_port_parse_error::HostPortParseError> for Error {
```
Comment (casey) [inline src/error.rs:184]: Is there a call site which is invoking this conversion?

File: src/error.rs [hunk 2]
```diff
@@ -167,12 +181,24 @@ impl From<globset::Error> for Error {
...
+  }
+}
+
 impl From<SystemTimeError> for Error {
   fn from(source: SystemTimeError) -> Self {
     Self::SystemTime { source }
   }
 }
 
+impl From<std::io::Error> for Error {
```
Comment (casey) [inline src/error.rs:196]: This should be removed, and more specific Io error variants added.

File: src/infohash.rs [hunk 3]
```diff
@@ -56,6 +56,12 @@ impl Infohash {
   }
 }
 
+impl Into<[u8; 20]> for Infohash {
```
Comment (casey) [inline src/infohash.rs:59]: I'm unsure I like this conversion, I feel like I'm not sure I want `.into()` conversions of `[u8; 20]` to Infohashes in the codebase, but I'm honestly not sure it's totally rational. It feels too general. It think maybe the conversion should be `Infohash::new(sha1hash)`

File: src/subcommand/torrent.rs [hunk 4]
```diff
@@ -14,6 +15,7 @@ mod verify;
   about("Subcommands related to the BitTorrent protocol.")
 )]
 pub(crate) enum Torrent {
+  Announce(announce::Announce),
```
Comment (casey) [inline src/subcommand/torrent.rs:18]: I'm not sure the name of the command should be `announce`. To me the connotation of `announce` is that I'm going to get other peers to connect to me, and doesn't so much have the connotation that we're just going to print out the peers that are present. I think `peers` is good, but I'm interested to hear your thoughts.

File: src/subcommand/torrent/announce.rs [hunk 5]
```diff
@@ -0,0 +1,135 @@
+use crate::common::*;
+
+const URI_HELP: &str = "Announce an infohash and list the response (a compact peer list).";
```
Comment (casey) [inline src/subcommand/torrent/announce.rs:3]: I think this should be more direct, like "Print torrent peers". Also, it should be named after the name of the command, so "PEERS_HELP". (Pending command name.)

File: src/subcommand/torrent/announce.rs [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+
+const INPUT_HELP: &str =
+  "Generate a compact peer list from a metainfo at `INPUT`. If `INPUT` is `-`, read \
```
Comment (casey) [inline src/subcommand/torrent/announce.rs:6]: "Read torrent metainfo from `INPUT`. If `INPUT`…"

File: src/subcommand/torrent/announce.rs [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+const INPUT_POSITIONAL: &str = "<INPUT>";
+
+#[derive(StructOpt)]
+#[structopt(
+  help_message(consts::HELP_MESSAGE),
+  version_message(consts::VERSION_MESSAGE),
+  about(URI_HELP)
+)]
+pub(crate) struct Announce {
+  #[structopt(
```
Comment (casey) [inline src/subcommand/torrent/announce.rs:20]: Thanks for replicating this input flag/positional stuff. It's annoying, but I think it makes the CLI nicer.

File: src/subcommand/torrent/announce.rs [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+  pub(crate) fn run(self, env: &mut Env, options: &Options) -> Result<(), Error> {
+    let target = xor_args(
+      "input_flag",
+      &self.input_flag,
+      "input_positional",
+      &self.input_positional,
+    )?;
+    let input = env.read(target)?;
+
+    let infohash = Infohash::from_input(&input)?;
```
Comment (casey) [inline src/subcommand/torrent/announce.rs:52]: Good catch not getting the `infohash` from the metainfo. I actually just [landed a PR](https://github.com/casey/intermodal/pull/468) that makes it more obvious that this shouldn't be done, and found a bug of mine in the process.

File: src/subcommand/torrent/announce.rs [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+    let metainfo = Metainfo::from_input(&input)?;
+
+    let mut rng = rand::thread_rng();
```
Comment (casey) [inline src/subcommand/torrent/announce.rs:55]: I think that `rng` should be part of env. This can be used to make RNG calls deterministic during testing, which seems useful. However, it's definitely something that should be done in a different PR, so don't worry about it. I opened #470 to track this.

File: src/subcommand/torrent/announce.rs [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+    let peer_id: [u8; 20] = rng.gen();
+    let mut peer_list = Vec::new();
+
+    if metainfo.trackers().peekable().peek().is_none() {
```
Comment (casey) [inline src/subcommand/torrent/announce.rs:59]: To get the trackers, you can do something like this up front:

```rust
let trackers = metainfo.trackers.collect::<Vec<Result<Url, Error>>>().filter_map(|result| {
  // print error message if tracker failed to parse
  // return None if failed to parse, so it gets filtered out
}).collect::<Vec<Url>>();

if trackers.is_empty() {
  return Err(Error::MissingTrackers);
}
```

Or, actually, to keep going if any trackers are bad, we can do:

File: src/subcommand/torrent/announce.rs [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+      if !options.quiet {
```
Comment (casey) [inline src/subcommand/torrent/announce.rs:60]: `Env::status` will print an error message, so this println shouldn't be necessary.

File: src/subcommand/torrent/announce.rs [hunk 12]
```diff
@@ -0,0 +1,135 @@
...
+        println!("Supplied metainfo specifies no trackers.");
+      }
+      return Err(Error::NoPeerSource);
+    }
+
+    if !options.quiet {
+      println!("[1/2] Announcing {} to trackers.", &infohash);
```
Comment (casey) [inline src/subcommand/torrent/announce.rs:67]: All printing to stderr and stdout should be done via the `env` object, so it can be captured when running tests. There are macros `err`, `errln`, `out`, and `outln` that make doing this convenient.

File: src/udp_tracker.rs [hunk 13]
```diff
@@ -0,0 +1,311 @@
...
+  // something readable and fast. Here are some options:
+  //
+  // 1. Cursor. With multiple write calls. ◆
+  //
+  // 2. IoSlice. Is there overhead to all the ::new() calls? This will probably be
+  //    the winner once the write_vectored() calls diffuse out to std::net
+  //
+  // 3. Serde. (Already a dependency). Seems like overkill.
+  //
+  // 4. Binread.
```
Comment (casey) [inline src/udp_tracker.rs:55]: I don't have any particularly good suggestions for which crate to use. Binread looks really cool, but doesn't support serialization. I think multiple write calls might be okay for now, if there isn't a clear winner otherwise.

**Review by casey — CHANGES_REQUESTED** [review]
I didn't get to reviewing the meat of the code, but it's getting late so I'm going to leave it at this for now.

If you check continuous integration, there are a bunch of clippy lints, you can see them locally if your run `cargo clippy --all` or `just clippy`.

**Comment by casey** [comment]
I also updated this branch with new commits from master, which I shouldn't have done, since it creates nasty merge commits. Feel free to rebase and force push.

File: src/subcommand/torrent/announce.rs (hunk 7 above)
Comment (atomgardner) [reply src/subcommand/torrent/announce.rs:20]: There's a simple solution that's implemented by at least one widely used tool: no positional arguments---everything gets a flag.

File: src/subcommand/torrent/announce.rs (hunk 10 above)
Comment (atomgardner) [reply src/subcommand/torrent/announce.rs:59]: Half your message didn't make it. How's the listing below?

```rust
let trackers = metainfo.trackers().collect::<Vec<Result<Url, Error>>>()
      .into_iter()
      .filter_map(|result| {
        match result {
          Ok(tracker) => Some(tracker),
          Err(err) => {
            if !options.quiet {
              out!(env, "{:?}", err).ok();
            }
            None
          }
        }}).collect::<Vec<Url>>();
```
It looks kinda gross to me. But maybe because I haven't drunk enough rusty kool-aid

File: src/subcommand/torrent/announce.rs (hunk 12 above)
Comment (atomgardner) [reply src/subcommand/torrent/announce.rs:67]: Any downside to adding something like this to avoid !quiet pattern?
```rust
  ($env:expr, $opt:expr, $fmt:expr, $($arg:tt)*) => {
    if !opt.quiet {
      write!($env.out_mut(), $fmt, $($arg)*).context(crate::error::Stdout)
    }
  };
```

File: src/udp_tracker.rs (hunk 13 above)
Comment (atomgardner) [reply src/udp_tracker.rs:55]: I still reckon vectored writes will eventually be the best way to do this.

File: src/error.rs [hunk 14]
```diff
@@ -129,6 +137,12 @@ pub(crate) enum Error {
   SymlinkRoot { root: PathBuf },
   #[snafu(display("Failed to retrieve system time: {}", source))]
   SystemTime { source: SystemTimeError },
+  #[snafu(display("The UDP tracker connection id has expired."))]
+  UdpConnectionIdExpired,
+  #[snafu(display("UDP tracker has failed to respond 9 times."))]
+  UdpConnectionExhaustedRetries,
+  #[snafu(display("UDP tracker response differed from expectations"))]
+  UdpTrackerBadResponse,
```
[the remark this answers is not in the document]
Comment (atomgardner) [reply src/error.rs:145]: The `UdpConnectionIdExpired` error is presently returned when a client with stale `connection_id` tries to send a message. Here's the relevant section of BEP15: 

> UDP connections / spoofing
> In the ideal case, only 2 packets would be necessary. However, it is possible to spoof the source address of a UDP packet. The tracker has to ensure this doesn't occur, so it calculates a value (connection_id) and sends it to the client. If the client spoofed it's source address, it won't receive this value (unless it's sniffing the network). The connection_id will then be send to the tracker again in packet 3. The tracker verifies the connection_id and ignores the request if it doesn't match. Connection IDs should not be guessable by the client. This is comparable to a TCP handshake and a syn cookie like approach can be used to storing the connection IDs on the tracker side. A connection ID can be used for multiple requests. A client can use a connection ID until one minute after it has received it. Trackers should accept the connection ID until two minutes after it has been send.

[the remark this answers is not in the document]
Comment (atomgardner) [reply src/error.rs:145]: This is only used in `udp_tracker.rs:send_and_retry_with_backoff()`. The timings look something like: 

    /\ connection id lasts 1 minute
    /\ for n \in [0,8], back off for 15 * 2**n seconds before trying to resend

Upshot: The connection id will need to be refreshed at some point during the retrying. And but refreshing `connection_id` requires sending another message.

So I think there's some potential for live locking (?) that is sidestepped by completely erroring out once the `connection_id` is found to be stale.

File: src/error.rs (hunk 1 above)
Comment (atomgardner) [reply src/error.rs:184]: Hmm---nope. It must've slipped in when I linearized the git history.

File: src/infohash.rs (hunk 3 above)
Comment (atomgardner) [reply src/infohash.rs:59]: I think the advice these days is to do `Impl From<Infohash> for [u8; 20]` and get the `into()` for free. That's possibly why you \~feel\~ weird?

File: src/subcommand/torrent.rs (hunk 4 above)
Comment (atomgardner) [reply src/subcommand/torrent.rs:18]: An announce tells the tracker ---heya, I want H; do you know anyone who's ever wanted H?--- & so clients aren't getting a peer list unless they `announce` that they want one.

Also, there's really not enough complexity here, I think, to justify translating the spec's language into userese.

File: src/error.rs [hunk 15]
```diff
@@ -30,6 +30,12 @@ pub(crate) enum Error {
   FileSearch { source: ignore::Error },
   #[snafu(display("Invalid glob: {}", source))]
   GlobParse { source: globset::Error },
+  #[snafu(display("Failed to parse host port"))]
+  HostPort {
+    source: HostPortParseError,
+  },
+  #[snafu(display("io error {}", source))]
```
[the remark this answers is not in the document]
Comment (atomgardner) [reply src/error.rs:37]: Yeah, this catchall is also grabbing the `write_all()` calls on cursors. Happy to change them to below?
 
    msg.write_all(&buf).context(MessageSerialize { src: io::Error })?;

**Comment by atomgardner** [comment]
The only problem I see with your high level thought: all of the big Linux distros use https trackers.

File: src/error.rs (hunk 15 above)
[the remark this answers is not in the document]
Comment (casey) [reply src/error.rs:37]: Sorry, I missed the cursor writes. I believe `Cursor::write*` functions are infallible, since it's backed by a vector. Can those be changed to calls to `Vec::extend_from_slice`, so avoid needing to use `?` or `.unwrap`?

File: src/error.rs (hunk 14 above)
[the remark this answers is not in the document]
Comment (casey) [reply src/error.rs:145]: That all makes sense. In that case I think `TrackerConnectionExpired` is a better name for the error, since conceptually it's the whole connection that's expiring, although that's tracked by the id.

File: src/infohash.rs (hunk 3 above)
Comment (casey) [reply src/infohash.rs:59]: I misread it as being a conversion from `[u8; 20]` to an `Infohash`. So nevermind my original objection. But yes, this should be `impl From<Infohash>`.

File: src/udp_tracker.rs [hunk 16]
```diff
@@ -0,0 +1,311 @@
...
+    }
+  }
+
+  pub fn announce(&self, btinh: Infohash) -> Result<Vec<SocketAddr>> {
+    let mut rng = rand::thread_rng();
+    let req = AnnounceRequest {
+      connection_id: self.get_connection_id()?,
+      action: 0x0001,
+      transaction_id: rng.gen(),
+      infohash: Infohash::into(btinh),
```
Comment (casey) [inline src/udp_tracker.rs:207]: This can be `btinh.into()`

File: src/subcommand/torrent.rs (hunk 4 above)
Comment (casey) [reply src/subcommand/torrent.rs:18]: I think discoverability is not so great, since it isn't immediately obvious you can get a list of peers from a command called `announce`. However, it does strike me that an `announce` command could be extended with other things, like the ability to set downloaded/left/uploaded bytes, etc, whereas such extensions wouldn't make sense with a `peers` command.

And, agreed, it's a good idea to avoid translating spec language. So let's go with announce.

File: src/subcommand/torrent/announce.rs (hunk 5 above)
Comment (casey) [reply src/subcommand/torrent/announce.rs:3]: Actually, if the command name is `announce` I think "Announce a torrent and print the returned peers." is better.

File: src/subcommand/torrent/announce.rs (hunk 7 above)
Comment (casey) [reply src/subcommand/torrent/announce.rs:20]: That's what I did initially, but users complained that it was too verbose.

File: src/subcommand/torrent/announce.rs (hunk 10 above)
Comment (casey) [reply src/subcommand/torrent/announce.rs:59]: Yah, that's a little gross. I think a for loop is better, since it can be tricky to use `?` inside a closure:

```rust
let mut trackers = Vec::new();
for result in metainfo.trackers() {
  match result {
    Ok(tracker) => trackers.push(tracker),
    Err(err) => if !options.quiet {
      outln!("Failed to parse tracker URL: {}", err)?;
    }
  }
}
```

It's totally fine to use a for loop. Iterator chains are ideal when they're a good fit, but when you've got to return errors in the middle I tend to use a for loop.

Comment (casey) [reply src/subcommand/torrent/announce.rs:59]: Updated the above to use `errln`, since it's a warning message.

File: src/subcommand/torrent/announce.rs (hunk 12 above)
Comment (casey) [reply src/subcommand/torrent/announce.rs:67]: Good idea! Actually, I think it might be better to just suppress all output to stderr if the `--quiet` flag is passed. This can be done centrally in `OutputStream`. I just opened #473 to track this. You can skip adding `if !quiet`, since this is simple enough that I should be able to add it soon.

File: src/udp_tracker.rs (hunk 13 above)
Comment (casey) [reply src/udp_tracker.rs:55]: I think at the moment we shouldn't worry about performance at all. With single announces, any performance improvements will be unnoticeable to end users. For simplicity, I'd just do extending vectors with byte slices, and then at the end write the whole vector. Then later if performance is an issue, we can profile and improve it.

**Comment by casey** [comment]
> The only problem I see with your high level thought: all of the big Linux distros use https trackers.

Sorry, I don't follow. If we add an HTTPS tracker later, can't the code be reorganized then?

File: src/subcommand/torrent/announce.rs (hunk 12 above)
Comment (casey) [reply src/subcommand/torrent/announce.rs:67]: This is done: #474.

**Comment by atomgardner** [comment]
> Sorry, I don't follow. If we add an HTTPS tracker later, can't the code be reorganized then?

It totally can. I was stating the obvious fact that if `imdl` is to become super serious software, loved by users everywhere, an http tracker client must be included.

**Comment by casey** [comment]
> It totally can. I was stating the obvious fact that if imdl is to become super serious software, loved by users everywhere, an http tracker client must be included.

100%!

**Comment by atomgardner** [comment]
This build pipeline is fastidious. Let's see if I can fix the book building.

**Comment by casey** [comment]
There's a lot of generated documentation, and a lot of checks that make sure that all the documentation is consistent and/or up-to-date.

I think in this case it's complaining that `imdl torrent announce` doesn't have an entry in `bin/gen/config.yaml`:

https://github.com/casey/intermodal/blob/master/bin/gen/config.yaml#L5

**Comment by atomgardner** [comment]
Interesting failure on the Windows machine. You totally saw it coming

**Comment by casey** [comment]
That's interesting! I'm surpised it would work on macos/linux, but fail on Windows.

**Comment by atomgardner** [comment]
I think Window is right to send back the error. Binding to `0.0.0.0:0` was definitely the wrong address.

**Comment by casey** [comment]
Ahh, yeah, I guess the tests should bind to 127.0.0.1?

File: bin/gen/config.yaml [hunk 17]
```diff
@@ -11,6 +11,10 @@ examples:
   text:    "BitTorrent metainfo related functionality is under the `torrent` subcommand:"
   code:    "imdl torrent --help"
 
+- command: imdl torrent announce
```
Comment (casey) [inline bin/gen/config.yaml:14]: I suspect that the primary initial use for this command will be to get the printed peer list, so that should be mentioned. Like, "Announce a torrent to its trackers and print all peers returned:"

File: src/error.rs [hunk 18]
```diff
@@ -30,6 +30,12 @@ pub(crate) enum Error {
   FileSearch { source: ignore::Error },
   #[snafu(display("Invalid glob: {}", source))]
   GlobParse { source: globset::Error },
+  #[snafu(display("Failed to parse host port"))]
```
[the remark this answers is not in the document]
Comment (casey) [reply src/error.rs:35]: This should print the source error, since those aren't automatically printed: `#[snafu(display("{}", source))]`

File: src/tracker/client.rs [hunk 19]
```diff
@@ -0,0 +1,268 @@
...
+    let mut buf = [0u8; 8192]; // is there a BUFSIZ macro for Rust?
+    let (resp, len) = self.roundtrip(&req, &mut buf)?;
+
+    if resp.transaction_id != req.transaction_id || resp.action != req.action {
+      return Err(Error::TrackerResponse);
+    }
+    self.parse_compact_peer_list(&buf[mem::size_of::<announce::Response>()..len])
+  }
+
+  fn roundtrip<T: Request>(&self, req: &T, rxbuf: &mut [u8]) -> Result<(T::Response, usize)> {
```
[the remark this answers is not in the document]
Comment (casey) [reply src/tracker/client.rs:74]: Instead of returning a `usize`, can `send_and_retry_with_backoff` and `roundtrip` return a `&[u8]`, which is the written-to portion of the rxbuf? This would avoid needing to index into the rxbuf correctly.

File: src/tracker/client.rs [hunk 20]
```diff
@@ -0,0 +1,268 @@
...
+    }
+
+    if len_read == 0 {
+      Err(Error::TrackerTimeout)
+    } else {
+      Ok(len_read)
+    }
+  }
+
+  // XXX: perhaps this should be in a different namespace
```
Comment (casey) [inline src/tracker/client.rs:115]: I think this is fine here.

File: src/tracker/client.rs [hunk 21]
```diff
@@ -0,0 +1,268 @@
...
+  fn parse_compact_peer_list(&self, buf: &[u8]) -> Result<Vec<SocketAddr>> {
+    let mut subswarm = Vec::<SocketAddr>::new();
+
+    let stride = if self.state.is_ipv6() { 18 } else { 6 };
+
+    // XXX: consider array_chunks()
```
Comment (casey) [inline src/tracker/client.rs:121]: Since array_chunks is nightly, and probably will be for a while, you can omit the note. And, actually, it wouldn't work here, since `stride` isn't a const.

File: src/subcommand/torrent/announce.rs [hunk 22]
```diff
@@ -0,0 +1,123 @@
...
+        Ok(tracker) => trackers.push(tracker),
+        Err(err) => errln!(env, "Failed to parse tracker URL: {}", err)?,
+      }
+    }
+
+    if trackers.is_empty() {
+      return Err(Error::MetainfoMissingTrackers);
+    }
+
+    let mut client = tracker::Client::new(peer_id)?;
```
[the remark this answers is not in the document]
Comment (atomgardner) [reply src/subcommand/torrent/announce.rs:67]: We didn't---we started with something tenuous like `the peer id's used in the peer handshake` and derived a bunch of maybes.

I like all of this (especially removing `State`). It'll also make it easier to solve the IPv6 problem I ran into when testing; namely, `client::new()` binds the UDP socket to a v4 address, so the `connect()` fails if the `HostPort::to_socket_addrs()` resolves into a v6 address.

File: src/subcommand/torrent/announce.rs [hunk 23]
```diff
@@ -0,0 +1,123 @@
...
+    errln!(env, "[1/2] Announcing {} to trackers.", &infohash)?;
+
+    for tracker in trackers {
+      if tracker.scheme() == "udp" {
```
[the remark this answers is not in the document]
Comment (atomgardner) [reply src/subcommand/torrent/announce.rs:71]: :100:

File: src/tracker/client.rs [hunk 24]
```diff
@@ -0,0 +1,268 @@
...
+      uploaded: 0x0000,
+      event: 0x0000,
+      ip_address: 0x0000,
+      num_want: u32::MAX,
+      port: self.sock.local_addr().context(error::Network)?.port(),
+    };
+    let mut buf = [0u8; 8192]; // is there a BUFSIZ macro for Rust?
+    let (resp, len) = self.roundtrip(&req, &mut buf)?;
+
+    if resp.transaction_id != req.transaction_id || resp.action != req.action {
```
[the remark this answers is not in the document]
Comment (atomgardner) [reply src/tracker/client.rs:68]: Sure can. But there'd be calls to getters instead, I think, since the function has the trait parameter.

File: src/error.rs [hunk 25]
```diff
@@ -44,23 +46,27 @@ pub(crate) enum Error {
...
-    input: InputTarget,
-    error: bendy::decoding::Error,
-  },
   #[snafu(display("Metainfo from {} failed to validate: {}", input, source))]
   MetainfoValidate {
     input: InputTarget,
     source: MetainfoError,
   },
+  #[snafu(display("Failed to interact with socket: {}", source))]
+  Network { source: io::Error },
```
[the remark this answers is not in the document]
Comment (atomgardner) [reply src/error.rs:69]: Sounds great

**Comment by atomgardner** [comment]
Yo!! I've had a bunch of changes queued up for a while now, and I think this is getting closer to usable. Let me know how it's looking.

I'm planning to push the Peer Protocol stuff up in the next few days.

**Comment by casey** [comment]
Awesome! It might take me a little bit, but I'll definitely get to this within the next few days.

**Review by casey — CHANGES_REQUESTED** [review]
10 days is a few, right 😅 Sorry for taking so long! I did some initial review, just really minor stuff, because I realized that I don't remember where we were in the feature.

What's new that I should be focusing on?

File: src/error.rs [hunk 26]
```diff
@@ -136,6 +142,19 @@ pub(crate) enum Error {
   SymlinkRoot { root: PathBuf },
   #[snafu(display("Failed to retrieve system time: {}", source))]
   SystemTime { source: SystemTimeError },
+  #[snafu(display("UDP tracker failed to resolve tracker `{}`: {}.", host_port, source))]
+  TrackerDNS {
+    host_port: HostPort,
+    source: io::Error,
+  },
+  #[snafu(display("UDP tracker connection id has expired."))]
+  TrackerConnectionExpired,
```
[the remark this answers is not in the document]
Comment (atomgardner) [reply src/error.rs:151]: It sure can

File: src/error.rs [hunk 27]
```diff
@@ -136,6 +142,19 @@ pub(crate) enum Error {
...
+  #[snafu(display("UDP tracker resolved `{}` to no useable addresses.", host_port))]
+  TrackerNoHosts { host_port: HostPort },
```
[the remark this answers is not in the document]
Comment (atomgardner) [reply src/error.rs:153]: My understanding of the DNS protocol is too shallow to know whether this behaviour is possible. It seems reasonable that whoever hatched  `to_socket_addrs` probably spent some serious time thinking it all through.

**Comment by atomgardner** [comment]
Heya heya---when 10 days is a few days plus holidays, you don't count the holidays.

I've pushed up your copy changes, but, I think, the "integration test" needs a look. And you should also try the command and confirm that it works.

**Review by casey — CHANGES_REQUESTED** [review]
Made a bunch of comments! Almost all of them requesting things that aren't tests. I created #490, to track things that don't belong in the initial PR.

I'm changing the tests to remove the one that generates the changelog, which means you also don't have to worry about adding a commit type to commits anymore. It was good for keeping the changelog up to date, but ultimately, I think, more annoying that it was worth.

If possible, could you push diffs instead of force pushing? This makes it easier to see what's changed since I last took a look. I think I might have said something about force pushing being preferable, but I take it all back.

File: src/subcommand/torrent/announce.rs [hunk 28]
```diff
@@ -0,0 +1,120 @@
...
+      match client.announce(infohash) {
+        Ok(subswarm) => {
+          peer_list.extend(subswarm);
+          break;
+        }
+        Err(err) => errln!(env, "{:?}", err)?,
+      }
+    }
+
+    for peer in &peer_list {
```
[the remark this answers is not in the document]
Comment (atomgardner) [reply src/subcommand/torrent/announce.rs:96]: There are a few tests in `src/tracker/client.rs` with the prefix `client_end_to_end_announce` that do this already, I think.

File: src/tracker/announce.rs [hunk 29]
```diff
@@ -0,0 +1,217 @@
...
+}
+
+impl Response {
+  pub(crate) const LENGTH: usize = 20;
+}
+
+impl super::Request for Request {
+  type Response = Response;
+
+  fn serialize(&self) -> Vec<u8> {
```
[the remark this answers is not in the document]
Comment (atomgardner) [reply src/tracker/announce.rs:40]: Awesome.

File: src/subcommand/torrent/announce.rs [hunk 30]
```diff
@@ -0,0 +1,120 @@
...
+      if let Err(e) = client {
+        errln!(env, "Couldn't connect to tracker: {}", e)?;
+        continue;
+      }
+      let client = client.invariant_unwrap("we know client.is_ok()");
+
+      match client.announce(infohash) {
+        Ok(subswarm) => {
+          peer_list.extend(subswarm);
+          break;
```
[the remark this answers is not in the document]
Comment (atomgardner) [reply src/subcommand/torrent/announce.rs:92]: Yeah, I like the idea of returning all peers. But the synchronous requesting will lead to this weird UX where after a peer list is dumped, the user's left waiting for the next list, etc...

File: src/tracker/client.rs (hunk 19 above)
[the remark this answers is not in the document]
Comment (atomgardner) [reply src/tracker/client.rs]: I'll do a little research on this. But it looks like I left myself a little comment about this:

```
src/client/response.rs:

pub(crate) trait Response {
  // We leak the response length so that payloads can be parsed. This should be
  // cleaner when associated types within traits can specify named lifetime
  // parameters.
  fn deserialize(buf: &[u8]) -> Result<(Self, &[u8])>

```

File: src/error.rs (hunk 27 above)
[the remark this answers is not in the document]
Comment (atomgardner) [reply src/error.rs:153]: Woah—looks like it works now!!

File: src/tracker/client.rs (hunk 19 above)
[the remark this answers is not in the document]
Comment (atomgardner) [reply src/tracker/client.rs]: Woah—looks like this is easily doable now!!

**Comment by casey** [comment]
I think this has gone stale 😅
